packages/core/elasticsearch/core-elasticsearch-client-server-internal/src/agent_manager.ts

```diff
@@ -0,0 +1,188 @@
+ * @internal
+ **/
+export class AgentManager {
+  /**
+   * Store Agent instances by type and protocol, e.g.:
+   *
+   * data: {
+   *   'http:': [agentInstance1, agentInstance2],
+   *   'https:': [agentInstance3, agentInstance4]
+   * },
+   * monitoring: {
+   *   'http:': [],
+   *   'https:': [agentInstance5]
+   * }
+   */
+  private agentStore: Record<string, Record<Protocol, Array<NetworkAgent | undefined>>>;
+  private logger: Logger;
+
+  constructor(
+    logger: Logger,
+    private defaultConfig: HttpAgentOptions = {
+      keepAlive: true,
+      keepAliveMsecs: 50000,
+      maxSockets: 256,
+      maxFreeSockets: 256,
+      scheduling: 'lifo',
+    }
+  ) {
+    this.logger = logger;
+    this.agentStore = {};
+  }
+
+  public getAgentFactory(
+    type: string,
+    agentOptions?: HttpAgentOptions | UndiciAgentOptions | AgentFactory | false
+  ): AgentFactory {
+    if (isAgentFactory(agentOptions)) {
+      // use the user-provided factory directly
+      return agentOptions;
+    }
+
+    const agentConfig = assertValidAgentConfig(agentOptions);
+
+    // a given agent factory (of a given type) always provides the same Agent instances (for the same protocol)
+    // we keep the indexes for each protocol, so that we can access them later on, when the factory is invoked
+    const { httpInstanceIndex, httpsInstanceIndex } = this.initTypeStore(type);
+
+    return (connectionOpts: ConnectionOptions): NetworkAgent => {
+      const agents = this.getAgentInstances(type, connectionOpts);
+
+      // when the factory is called for a given connection (and protocol), we use the appropriate index
+      const agentIndex = isHttps(connectionOpts) ? httpsInstanceIndex : httpInstanceIndex;
+
+      let agent = agents[agentIndex];
+
+      if (!agent) {
+        agent = this.createAgent(agentConfig, connectionOpts);
+
+        // Allow GC'ing the Agent after it has been terminated
+        const doDestroy = agent.destroy.bind(agent);
+        agent.destroy = () => {
+          doDestroy();
+          agents[agentIndex] = undefined;
+        };
+
+        // add the new Agent instance to the list of instances of that type
+        agents[agentIndex] = agent;
+      }
+
+      return agent;
+    };
+  }
+
+  private initTypeStore(type: string): { httpInstanceIndex: number; httpsInstanceIndex: number } {
+    let agentsOfType = this.agentStore[type];
+
+    if (!agentsOfType) {
+      agentsOfType = {
+        [HTTP]: [],
+        [HTTPS]: [],
+      };
+
+      this.agentStore[type] = agentsOfType;
+    } else {
+      this.logger.warn(
+        `An Agent factory for type ${type} has already been retrieved.
+        This is not optimal since the existing instance cannot be reused.
+        Please consider reusing the previously retrieved factory instead`
+      );
+    }
+
+    const httpInstanceIndex = agentsOfType[HTTP].length;
+    const httpsInstanceIndex = agentsOfType[HTTPS].length;
```

Comment:
Is there a scenario where these two indices would be different, given we're creating the entry for both protocols every time?

Reply:
No we're not, we're creating the entry for one or the other protocol depending on the `elasticsearch.hosts` configuration. We could be connecting to 2 nodes through HTTPS and to one node through HTTP for instance.

Reply:
> We could be connecting to 2 nodes through HTTPS and to one node through HTTP for instance.

Still, in the implementation you're creating one entry per call for a given `type` in both arrays, isn't that correct? so atm, the length of both is strictly equal, unless I missed something in the impl?

Reply:
The comment is outdated now, but FWIW I was pushing the newly created instances to either the http array or the https array, but not both at the same time.

Now I've simplified the store structure, but we still have the 2 arrays, and the code pushes to one or to the other, but not to both. So we could end up having 3 Agents of type Https and 1 of type Http.

The idea is that we provide the `AgentFactory` to the ES Client class, who loops through all the `elasticsearch.hosts` and creates an `HttpConnection` for each of them. 
Internally, each `HttpConnection` calls the `agentFn` to obtain an instance of Agent.
Then, depending on the protocol of `elasticsearch.hosts[i]`, the `AgentFactory` will provide an instance of `HttpAgent` or one of `HttpsAgent`, and it will store this instance in the appropriate queue in the manager.

Finally, note that for hosts using the same protocol the same instance will be provided.